src/Compilers/CSharp/Portable/Binder/SwitchBinder.cs

```diff
@@ -27,90 +29,102 @@ protected SwitchBinder(Binder next, SwitchStatementSyntax switchSyntax)
 
-            var map = new Dictionary<object, List<SourceLabelSymbol>>(labels.Length, new SwitchConstantValueHelper.SwitchLabelsComparer());
+            var map = new Dictionary<object, SourceLabelSymbol>(labels.Length, new SwitchConstantValueHelper.SwitchLabelsComparer());
             foreach (SourceLabelSymbol label in labels)
             {
                 SyntaxKind labelKind = label.IdentifierNodeOrToken.Kind();
+                if (labelKind == SyntaxKind.IdentifierToken) continue;
 
-                if (labelKind == SyntaxKind.CaseSwitchLabel ||
-                    labelKind == SyntaxKind.DefaultSwitchLabel)
+                object key;
+                var constantValue = label.SwitchCaseLabelConstant;
+                if (constantValue != (object)null && !constantValue.IsBad)
                 {
-                    object key;
-                    var constantValue = label.SwitchCaseLabelConstant;
-                    if (constantValue != null)
-                    {
-                        // Case labels with a non-null constant value are indexed on their ConstantValue.
-                        key = constantValue;
-                    }
-                    else if (labelKind == SyntaxKind.DefaultSwitchLabel)
-                    {
-                        // Default label(s) are indexed on a special DefaultKey object.
-                        key = s_defaultKey;
-                    }
-                    else
-                    {
-                        // Invalid case labels with null constant value are indexed on the labelName.
-                        key = label.Name;
-                    }
-
-                    List<SourceLabelSymbol> labelsList;
-                    if (!map.TryGetValue(key, out labelsList))
-                    {
-                        labelsList = new List<SourceLabelSymbol>();
-                        map.Add(key, labelsList);
-                    }
+                    // Case labels with a non-null constant value are indexed on their ConstantValue.
+                    key = KeyForConstant(constantValue);
+                }
+                else if (labelKind == SyntaxKind.DefaultSwitchLabel)
+                {
+                    // Default label(s) are indexed on a special DefaultKey object.
+                    key = s_defaultKey;
+                }
+                else
+                {
+                    // Invalid case labels with null constant value are indexed on the labelName.
+                    key = label.IdentifierNodeOrToken.AsNode();
+                }
 
-                    Debug.Assert(!labelsList.Contains(label));
-                    labelsList.Add(label);
+                if (!map.ContainsKey(key))
+                {
+                    map.Add(key, label);
+                }
+                else
+                {
+                    // If there is a duplicate label, ignore it. It will be reported when binding the switch label.
```

Comment:
Will change it.

---
In reply to: [70175251](https://github.com/dotnet/roslyn/pull/12416#discussion_r70175251) [](ancestors = 70175251)